### ISSUE TITLE
ONNX package is unnecessarily large

### ISSUE THREAD

**Issue by prasanthpul** [onnx/onnx#2917]
The ONNX whl package for Linux is 7.4 MB. This is 2x the size of a more complex package like onnxruntime which is surprising since ONNX code is very small and minimal. The package seems to contain a lot of test data which is bloating the package. The scripts to generate test data are in the package. Not everyone needs the test data, so we should remove the test data from the package and users who need it can use the included scripts to generate it after pip install.

**Comment by snnn** [comment]
I agree with you, but, the test data in the package are not the same as the things auto-generated, especially opset version often mismatches.   Auto-generated ones always use the latest opset, which I think is better.

**Comment by prasanthpul** [comment]
@snnn sounds like you are saying the test data in the package is either incomplete or outdated. In either case, sounds like it may not be as valuable as believed. We should fix that.

The generation code should be made to produce data in different opsets. Alternatively, provide pre-generated datasets for each opset as separate downloadable artifacts on Github.

**Comment by snnn** [comment]
> The generation code should be made to produce data in different opsets. Alternatively, provide pre-generated datasets for each opset as separate downloadable artifacts on Github.

Agree. See the discussion in #2912. 

Onnxruntime has a strong need on this.

**Comment by askhade** [comment]
> > The generation code should be made to produce data in different opsets. Alternatively, provide pre-generated datasets for each opset as separate downloadable artifacts on Github.
> 
> Agree. See the discussion in #2912.
> 
> Onnxruntime has a strong need on this.

I like the option to provide pre-generated datasets as github artifacts... 
generation code creates the test data based on the node test cases. Everytime an op is updated we update the test case to reflect the changes... If we were to support generation of test data for all opsets then we would have to maintain the test cases for every op for every opset. This is huge and will continue growing.

**Comment by askhade** [comment]
@snnn Today onnxruntime solves this problem by installing every version of onnx in the test pipelines... What are the cons of this approach? And onnx provides this old opset test data as a github artifact will it solve these cons?

**Comment by askhade** [comment]
> I agree with you, but, the test data in the package are not the same as the things auto-generated, especially opset version often mismatches. Auto-generated ones always use the latest opset, which I think is better.

This is because the testdata is not regenerated to use latest opset for every test case. It is regenerated case by case only when a node test is updated ( which only happens when an op is updated). Therefore for example an op which was last updated in opset 7 will have the model in test data use opset7... It also means even if the data is regenerated 99% there wont be any issue because this op has not been updated. In 1% cases there can be an issue if the test case also uses another op which was updated but this is very very rare... these node test are unit tests for operators and do not contain other ops...

**Comment by snnn** [comment]
> @snnn Today onnxruntime solves this problem by installing every version of onnx in the test pipelines... What are the cons of this approach? 

Some old ONNX releases are not installable anymore. Onnxruntime need some hacks to make it work. We were bitten by it twice, in the last two ORT releases. For example, we just spent one week on this when preparing the ORT 1.4 release. We don't want to continue to maintain these hacks. And, now Onnxruntime compile all ONNX version from source code for every new pull request, it's costly. We can bear it but we can't recommend it to the others. 

> And onnx provides this old opset test data as a github artifact will it solve these cons?

It's ok. My only worries is: For many times, the pre-generated test data in ONNX repo doesn't match the scripts they were from. It happened a lots when we were preparing the ONNX 1.7 release, we spent one month to fix it but even that the final release still doesn't seem correct.  But I think it's fine, it can be improved. 

Another thing: as you can see, currently we have some pytorch converted test data in the latest ONNX, however, these data are old, I guess nobody knows how these data were generated. And, for sure, they're not perfect. There are some errors we want to fix but it's not that easy. 

I think scripts are more user friendly than the generated data. For example, if ORT couldn't pass one of the tests, if we have the script, we at least can know how the ground truth were from. Otherwise, it's harder to debug and make it test pass.

**Comment by snnn** [comment]
> This is because the testdata is not regenerated to use latest opset for every test case. It is regenerated case by case only when a node test is updated ( which only happens when an op is updated). Therefore for example an op which was last updated in opset 7 will have the model in test data use opset7... It also means even if the data is regenerated 99% there wont be any issue because this op has not been updated. In 1% cases there can be an issue if the test case also uses another op which was updated but this is very very rare... these node test are unit tests for operators and do not contain other ops...

Right, but they're still different. It's often that ORT passed one of them but failed on the other.

**Comment by askhade** [comment]
I think we all agree that the package itself should not contain the data... It already contains the cases and the script to generate the data from it... We should definitely strip the data from the package and if there is any existing bug which prevents us from doing this then we should fix it... @prasanthpul what should be the next step here? get approval from arch-infra sig?

Regarding the test data for all opset versions there are 2 options discussed
1. Add test cases for every version of the op
2. Add the created data for the older opset version as github artifact which backends can download

Let's discuss these 2 options in the upcoming Operator SIG meeting which is on July 24
Details : https://gitter.im/onnx/operators


@ebarsoum @postrational FYI

**Comment by prasanthpul** [comment]
@askhade yeah, the proposal should be made in Arch/Infra SIG and decided. 
cc @linkerzhang to tell when the next SIG mtg is

**Comment by Korijn** [comment]
As an alternative, you could also leave the test data for all opsets in the repo, so consumers can simply use git (shallow) clone of specific subfolders of the repo to pull the data.

**Comment by postrational** [comment]
I would insist that the current package content should be installable from PyPI. However, the package could be split into two and the one package could contain only core functionality, another test data.

Small core could be installed using:

    pip install onnx

The complete package:

    pip install onnx[tests]

**Comment by askhade** [comment]
@postrational : 2 different packages adds a lot of maintenance overhead. The release process is already pretty heavy and adding another package will only add to the complexity. What are your concerns for generating the data from script provided the script and test cases are part of the onnx package?
